src/main/java/org/sigmah/server/service/ContactHistoryService.java

```diff
@@ -84,62 +91,93 @@ public int compare(ContactHistory first, ContactHistory second) {
+        contactHistory.setId(historyToken.getId());
+        contactHistory.setComment(historyToken.getComment());
+        contactHistory.setUpdatedAt(historyToken.getDate());
+
+        if (historyToken.getUser() != null) {
+          contactHistory.setUserFullName(historyToken.getUser().getFullName());
+        }
+        if (flexibleElement instanceof ContactListElement) {
+          contactHistory.setFormattedChangeType(getRelationshipFormatter("Contact", historyToken.getType(), language));
+          Contact relatedContact = contactDAO.findById(Integer.parseInt(historyToken.getValue()));
+          contactHistory.setSubject(relatedContact.getFullName());
+          contactHistory.setFormattedValue(flexibleElement.getLabel());
+          contactHistory.setValueType(ContactHistory.ValueType.STRING);
+        } else {
+          contactHistory.setFormattedChangeType(i18nServer.t(language, "contactHistoryChangeTypeUpdatedProperty"));
+          contactHistory.setFormattedValue(modelPropertyService.getFormattedValue(flexibleElement, historyToken.getValue(), language));
+
+          if (flexibleElement instanceof DefaultContactFlexibleElement) {
+            contactHistory.setSubject(modelPropertyService.getDefaultContactPropertyLabel(((DefaultContactFlexibleElement) flexibleElement).getType(), language));
+
+            if (((DefaultContactFlexibleElement) flexibleElement).getType() == DefaultContactFlexibleElementType.PHOTO) {
+              contactHistory.setValueType(ContactHistory.ValueType.IMAGE);
             } else {
-              contactHistory.setSubject(flexibleElement.getLabel());
               contactHistory.setValueType(ContactHistory.ValueType.STRING);
             }
+          } else {
+            contactHistory.setSubject(flexibleElement.getLabel());
+            contactHistory.setValueType(ContactHistory.ValueType.STRING);
           }
-          contactHistories.add(contactHistory);
         }
+        contactHistories.add(contactHistory);
       }
     }
     return contactHistories;
   }
 
-  private List<ContactHistory> generateHistoryFromInboundRelationships(Contact contact, Language language) {
+  private List<ContactHistory> generateHistoryFromInboundRelationships(Contact contact, Language language, boolean lastOnly) {
     List<ContactHistory> contactHistories = new ArrayList<ContactHistory>();
-    List<HistoryToken> historyTokens = historyTokenDAO.findByIdInSerializedValue(contact.getId());
+    List<HistoryToken> historyTokens = historyTokenDAO.findByIdInSerializedValueAndElementType(contact.getId(), EntityConstants.CONTACT_LIST_ELEMENT_TABLE, lastOnly);
     for (HistoryToken historyToken : historyTokens) {
       FlexibleElement flexibleElement = flexibleElementDAO.findById(historyToken.getElementId());
       // We are looking here for relationships going on the current contact
       // The flexible element must be a contact list element
+      // This should have already been filtered by the request.
```

Comment:
Instead of a comment, could you thow an `IllegalStateException` instead of just `continue`? This way, we may not ignore a bug in `findByIdInSerializedValueAndElementType`.